### PR TITLE
adding vmss extension

### PR DESCRIPTION
**Pull request by zikalino**
##### SUMMARY
This module adds VMSS Extension functionality, which is quite essential when using VMSS.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
azure_rm_virtualmachinescalesetextension

### CODE REVIEW

**Comment by Fred-sun** [comment]
@yungezz @yuwzho

**Review by yungezz — APPROVED** [review]

**Comment by Fred-sun** [comment]
@nitzmahone Please help review this PR  and push for merge.  Thanks!

File: lib/ansible/modules/cloud/azure/azure_rm_virtualmachinescalesetextension.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+            - A dictionary containing protected extension settings.
+            - Settings depend on extension type.
+            - Refer to U(https://docs.microsoft.com/en-us/azure/virtual-machines/extensions/overview) for more information.
+    auto_upgrade_minor_version:
+        description:
+            - Whether the extension handler should be automatically upgraded across minor versions.
+        type: bool
+    state:
+        description:
+            - Assert the state of the SQL server. Use 'present' to create or update a server and
```
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_virtualmachinescalesetextension.py]: s/SQL server/VMSS extension/

File: lib/ansible/modules/cloud/azure/azure_rm_virtualmachinescalesetextension.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+
+    def create_or_update_vmssextension(self):
+        self.log("Creating VMSS extension {0}".format(self.name))
+        try:
+            params = self.compute_models.VirtualMachineScaleSetExtension(
+                location=self.location,
+                publisher=self.publisher,
+                type=self.type,
+                type_handler_version=self.type_handler_version,
+                auto_upgrade_minor_version=self.auto_upgrade_minor_version,
```
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_virtualmachinescalesetextension.py:260]: Does the API handle `None` OK here, or do we need a default?

File: lib/ansible/modules/cloud/azure/azure_rm_virtualmachinescalesetextension.py [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+                else:
+                    self.settings = response.get('settings')
+
+                if self.protected_settings is not None:
+                    if response.get('protected_settings') != self.protected_settings:
+                        response['protected_settings'] = self.protected_settings
+                        to_be_updated = True
+                else:
+                    self.protected_settings = response.get('protected_settings')
+
```
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_virtualmachinescalesetextension.py:218]: Are location/version/type/auto-upgrade things that should be checked for changes? Not sure which (if any) are updateable...

**Review by nitzmahone — CHANGES_REQUESTED** [review]
Couple minor things and outstanding questions, but generally looks good.

File: lib/ansible/modules/cloud/azure/azure_rm_virtualmachinescalesetextension.py (hunk 3 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_virtualmachinescalesetextension.py:218]: yeah... I will check all of those. there should be warnings if not updatable (according to our recent discussion).
actually vmss extension is very similar to vm extension, so I will update both if necessary.

File: lib/ansible/modules/cloud/azure/azure_rm_virtualmachinescalesetextension.py (hunk 2 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_virtualmachinescalesetextension.py:260]: yes it does

**Comment by zikalino** [comment]
@nitzmahone i have just revisited azure_virtualmachine_extension as well.
fixed all the idempotence issues according to our current guidelines:
https://github.com/ansible/ansible/pull/51005
i have applied the same fixes here.

**Comment by mattclay** [comment]
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/102624/77/tests

**Review by nitzmahone — APPROVED** [review]

**Pull request merged**